### PR TITLE
start on logic chapter; universal quantifiers and implication

### PR DESCRIPTION
**Pull request by avigad**

### CODE REVIEW

File: source/logic.rst [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+    #check ∀ x y ε : ℝ,
+      0 < ε → ε ≤ 1 → abs x < ε → abs y < ε → abs (x * y) < ε
+    -- END
+
+In words, we would say "for every ``x``, ``y``, and ``ε``,
+if ``0 < ε ≤ 1``, the absolute value of ``x`` is less than ``ε``,
+and the absolute value of ``y`` is less than ``ε``,
+then the absolute value of ``x * y`` is less than ``ε``."
+In Lean, the implication arrows associate to the right,
+making it easy to chain implications in this way.
```
Comment (PatrickMassot) [inline source/logic.rst:38]: Mathematicians don't know what "associates to the right means". We need to have an example like: "P → Q → R" means 'P → (Q → R)". And even with parentheses this is still hard to parse for regular mathematicians.

File: source/logic.rst [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    lemma my_lemma : ∀ x y ε : ℝ,
+      0 < ε → ε ≤ 1 → abs x < ε → abs y < ε → abs (x * y) < ε :=
+    sorry
+
+    section
```
Comment (PatrickMassot) [inline source/logic.rst:56]: Did we explain what section does?

File: source/logic.rst [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+      0 < ε → ε ≤ 1 → abs x < ε → abs y < ε → abs (x * y) < ε :=
+    begin
+      intros x y ε epos ele1 xlt ylt,
+      sorry
+    end
+    -- END
+
+We can use any names we want for the universally quantified variables;
+they do not have to be ``x``, ``y``, and ``ε``.
+(If there is only one variable, you can use the singular ``intro``
```
Comment (PatrickMassot) [inline source/logic.rst:115]: I wouldn't insist on that singular feature. I think it was a very bad idea. I've seen a lot of students struggling because of the existence of `intro`, especially since the error message is not helpful.

File: source/logic.rst [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+    end
+    -- END
+
+At this stage, you also know that if you use
+the ``apply`` tactic to apply ``my_lemma``
+to a goal of the form ``abs (a * b) < δ``,
+you are left with new goals that require you to  prove
+each of the hypotheses.
+
+To prove a statement like this, use the ``intros`` tactic.
```
Comment (PatrickMassot) [inline source/logic.rst:97]: Maybe mention sometimes that we still have to introduce `x y ε` even if they are implicit. Implicitness is about using the lemma, it doesn't change anything to the proof.

File: source/logic.rst [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+.. code-block:: lean
+
+    import data.real.basic
+
+    -- BEGIN
+    lemma my_lemma : ∀ {x y ε : ℝ},
+      0 < ε → ε ≤ 1 → abs x < ε → abs y < ε → abs (x * y) < ε :=
+    begin
+      intros x y ε epos ele1 xlt ylt,
+      suffices h : abs x * abs y < 1 * ε,
```
Comment (PatrickMassot) [inline source/logic.rst:135]: I don't think this is a great example. The intermediate statement is not something we would state on paper, and there is no issue using `calc` from start to finish. I'm sure we can find more natural places to introduce `suffices`.

File: source/logic.rst [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+      apply hgb
+    end
+    -- END
+
+Applying ``intro`` to the predicate ``fn_ub (λ x, f x + g x) (a + b)``
+forces Lean to unfold the definition of ``fn_ub``
+and introduce ``x`` for the universal quantifier.
+The goal is then ``(λ (x : ℝ), f x + g x) x ≤ a + b``.
+But applying ``(λ x, f x + g x)`` to ``x`` should result in ``f x + g x``,
+and the ``dsimp`` command performs that simplification.
```
Comment (PatrickMassot) [inline source/logic.rst:206]: Did you already explain `change`? You could mention it here, as it offers much more control.

File: source/logic.rst [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+of any structure that is an "ordered additive commutative monoid";
+the details of what that means don't matter now,
+but it is worth knowing that the natural numbers, integers, rationals,
+and real numbers are all instances.
+So if we prove the theorem ``fn_ub_add`` at that level of generality,
+it will apply in all these instances.
+
+.. code-block:: lean
+
+    variables {α : Type*} {R : Type*} [ordered_cancel_add_comm_monoid R]
```
Comment (PatrickMassot) [inline source/logic.rst:258]: Maybe remind our readers they already met square brackets in Section 2.2, although the actual explanation is still to come.

File: source/logic.rst [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+    import data.real.basic
+
+    variables (f g : ℝ → ℝ)
+
+    -- BEGIN
+    def even (f : ℝ → ℝ) : Prop := ∀ x, f x = f (-x)
+    def odd (f : ℝ → ℝ) : Prop := ∀ x, f x = - f (-x)
+
+    example (ef : even f) (eg : even g) : even (λ x, f x + g x) :=
+    by { intro x, dsimp, rw [ef, eg] }
```
Comment (PatrickMassot) [inline source/logic.rst:380]: Should we mention that this `dsimp` (or some `change`) is *needed* here because of `rw`. This is also a place where a `calc` block would read much nicer, including an explicit unfolding line:
```lean
λ x,
calc (λ x, f x + g x) x = f x + g x : rfl
... = f (-x) + g (-x) : by rw [ef, eg]
```
or the same in tactic mode of course.

File: source/logic.rst [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+
+    import data.real.basic
+
+    -- BEGIN
+    open function
+
+    example (c : ℝ) : injective (λ x, x + c) :=
+    begin
+      intros x₁ x₂ h',
+      apply eq_of_add_eq_add_right h',
```
Comment (PatrickMassot) [inline source/logic.rst:467]: This proof feels like cheating. You could have written the full proof term `λ _ _, eq_of_add_eq_add_right`. And our readers will want to know how to subtract `c` from both side of `h'` and simplify. If we are not ready to explain how to do that then maybe we should choose another example (such that injectivity of a composition of injective functions).

File: source/logic.rst [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+    end
+
+    example {c : ℝ} (h : c ≠ 0) : injective (λ x, c * x) :=
+    sorry
+    -- END
+
+Finally, show that the composition of two injective functions is injective:
```
Comment (PatrickMassot) [inline source/logic.rst:474]: Oh, that was the next example :-)

File: source/logic.rst (hunk 6 above)
Comment (avigad) [reply source/logic.rst:206]: I think I used the syntactic equivalent `show` in the previous chapter. Can I stick with `show`?

File: source/logic.rst (hunk 9 above)
Comment (avigad) [reply source/logic.rst:467]: I agree, it is not a great example. I wanted readers to look at what happens after the intros. I'll see if I can think of a better one (but maybe I'll leave it since you like the composition one, that comes next).

File: source/logic.rst (hunk 6 above)
Comment (PatrickMassot) [reply source/logic.rst:206]: The issue with `show` is it works only with the goal, whereas you can `change ... at ...`.

File: source/logic.rst (hunk 2 above)
Comment (avigad) [reply source/logic.rst:56]: Yes, in the previous chapter, "We can delimit the scope of the declaration by putting it
in a ``section ... end`` block." But if you think I should remind readers, let me know, and I will.
